src/core/execute.c

```diff
@@ -4523,224 +4524,413 @@ int exec_command_append(ExecCommand *c, const char *path, ...) {
         return 0;
 }
 
+static void *remove_tmpdir_thread(void *p) {
+        _cleanup_free_ char *path = p;
 
-static int exec_runtime_allocate(ExecRuntime **rt) {
+        (void) rm_rf(path, REMOVE_ROOT|REMOVE_PHYSICAL);
+        return NULL;
+}
+
+static ExecRuntime* exec_runtime_free(ExecRuntime *rt, bool destroy) {
+        int r;
+
+        if (!rt)
+                return NULL;
 
+        if (rt->manager)
+                (void) hashmap_remove(rt->manager->exec_runtime_by_id, rt->id);
+
+        /* When destroy is true, then rm_rf tmp_dir and var_tmp_dir. */
+        if (destroy && rt->tmp_dir) {
+                log_debug("Spawning thread to nuke %s", rt->tmp_dir);
+
+                r = asynchronous_job(remove_tmpdir_thread, rt->tmp_dir);
+                if (r < 0) {
+                        log_warning_errno(r, "Failed to nuke %s: %m", rt->tmp_dir);
+                        free(rt->tmp_dir);
+                }
+
+                rt->tmp_dir = NULL;
+        }
+
+        if (destroy && rt->var_tmp_dir) {
+                log_debug("Spawning thread to nuke %s", rt->var_tmp_dir);
+
+                r = asynchronous_job(remove_tmpdir_thread, rt->var_tmp_dir);
+                if (r < 0) {
+                        log_warning_errno(r, "Failed to nuke %s: %m", rt->var_tmp_dir);
+                        free(rt->var_tmp_dir);
+                }
+
+                rt->var_tmp_dir = NULL;
+        }
+
+        rt->id = mfree(rt->id);
+        rt->tmp_dir = mfree(rt->tmp_dir);
+        rt->var_tmp_dir = mfree(rt->var_tmp_dir);
+        safe_close_pair(rt->netns_storage_socket);
+        return mfree(rt);
+}
+
+static void exec_runtime_freep(ExecRuntime **rt) {
         if (*rt)
-                return 0;
+                (void) exec_runtime_free(*rt, false);
+}
+
+static int exec_runtime_allocate(ExecRuntime **rt) {
+        assert(rt);
 
         *rt = new0(ExecRuntime, 1);
         if (!*rt)
                 return -ENOMEM;
 
-        (*rt)->n_ref = 1;
         (*rt)->netns_storage_socket[0] = (*rt)->netns_storage_socket[1] = -1;
-
         return 0;
 }
 
-int exec_runtime_make(ExecRuntime **rt, ExecContext *c, const char *id) {
+static int exec_runtime_add(
+                Manager *m,
+                const char *id,
+                const char *tmp_dir,
+                const char *var_tmp_dir,
+                int netns_storage_socket[2],
+                ExecRuntime **ret) {
+
+        _cleanup_(exec_runtime_freep) ExecRuntime *rt = NULL;
         int r;
 
-        assert(rt);
-        assert(c);
+        assert(m);
         assert(id);
 
-        if (*rt)
-                return 1;
+        r = hashmap_ensure_allocated(&m->exec_runtime_by_id, &string_hash_ops);
+        if (r < 0)
+                return r;
```

Comment:
There's an inconsistency here: all checks below use log_oom(), but this one is silent. I think it'd be OK to make this function not log at all and require the callers to log if necessary.

Reply:
Indeed. Fixed. Thank you.